File: abc_training/C/C_Invisible_Hand.rs
//            C - Invisible Hand
// ----------------------------------------
// 問題
// https://atcoder.jp/contests/abc312/tasks/abc312_c
// ----------------------------------------

// attributes
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(non_snake_case)]
#![allow(dead_code)]
#![allow(unused_macros)]

use std::collections::BTreeMap;

// imports
use itertools::Itertools;
use proconio::{
    fastout, input,
    marker::{Bytes, Chars, Usize1},
};
use superslice::Ext;

macro_rules! debug {
    ( $($val:expr),* $(,)* ) => {{
        #[cfg(debug_assertions)]
        eprintln!( concat!($(stringify!($val), " = {:?}, "),*), $($val),* );
    }};
}

const INF: usize = 1001001001;

// main
fn main() {
    input! {
        N: usize,
        M: usize,
        mut A: [usize; N],
        mut B: [usize; M],
    }

    A.sort();
    B.sort();

    // 累積和
    let mut SA = vec![0; N + 1];
    for (i, &a) in A.iter().enumerate() {
        SA[i + 1] = SA[i] + a;
    }

    let mut SB = vec![0; M + 1];
    for (i, &b) in B.iter().enumerate() {
        SB[i + 1] = SB[i] + b;
    }

    // 二分探索
    let mut ng = 0;
    let mut ok = INF;

    while ok - ng > 1 {
        let mid = (ng + ok) / 2;
        // 座標をもどす
        let ma = A.lower_bound(&mid);
        let mb = B.lower_bound(&mid);

    }

    println!("{}", ok);
}

File: contests/abc312/C_Invisible_Hand.py
#            C - Invisible Hand           
# ----------------------------------------
# 問題
# https://atcoder.jp/contests/abc312/tasks/abc312_c
# ----------------------------------------

from bisect import bisect_left, bisect_right

N, M = map(int, input().split())
A = sorted(map(int, input().split()))
B = sorted(map(int, input().split()))

ng = 0
ok = 1e20

while ok - ng > 1:
    mid = (ng + ok) // 2

    # Aのmid未満の要素
    sell = bisect_right(A, mid)

    # Bのmid以上の要素
    buy = M - bisect_left(B, mid)

    if sell >= buy:
        ok = mid
    else:
        ng = mid

print(int(ok))
